File: projects/flock02/src/Vector.java
public class Vector {
    double x, y;
    
    public double getX() { return x; }
    public double getY() { return y; }
    public void setX(double x) { this.x = x; }
    public void setY(double y) { this.y = y; }
    
    public Vector(double x, double y) {
        this.x = x;
        this.y = y;
    }
    
    public double magnitude() {
        return Math.sqrt((x * x) + (y * y));
    }
    public void scale(double n) {
        x *= n;
        y *= n;
    }
    public void normalize() {
        double magnitude = magnitude();
        if(magnitude == 0) {
            scale(0);
        }
        else {
            scale(1/magnitude());
        }
    }
    public void add(Vector v) {
        x += v.x;
        y += v.y;
    }
    public void subtract(Vector v) {
        x -= v.x;
        y -= v.y;
    }
    public void dot(Vector v) {
        x *= v.x;
        y *= v.y;
    }
    public void cross(Vector v) {
        x *= v.y;
        y *= v.x;
    }
    public String toString() {
        return VectorMath.getHeading(this) + " degrees for " + magnitude() + " units; [" + x + ", " + y + "]";
    }
    public Object clone() {
        return new Vector(x, y);
    }
}

File: projects/flockpainter/vector.js
function Vector(x, y) {
    this.x = x || 0;
    this.y = y || 0;
}

Vector.prototype.magnitude = function() {
    return Math.sqrt(this.x * this.x + this.y * this.y);
};
Vector.prototype.toString = function() {
    return '(' + this.x + ',' + this.y + ')';
};

// Vectors as immutable objects
Vector.prototype.scale = function(n) {
    return new Vector(this.x * n, this.y * n);
};
Vector.prototype.add = function(v) {
    return new Vector(this.x + v.x, this.y + v.y);
};
Vector.prototype.normalize = function() {
    var m = this.magnitude();
    return (m === 0) ? this.scale(0) : this.scale(1/m);
};

// In-place operations
Vector.prototype.iscale = function(n) {
    this.x *= n;
    this.y *= n;
};
Vector.prototype.iadd = function(v) {
    this.x += v.x;
    this.y += v.y;
};
Vector.prototype.inormalize = function() {
    var m = this.magnitude();
    this.iscale((m === 0) ? 0 : 1/m);
};
